### ISSUE TITLE
Vim9: cannot force folds recomputation by evaluating folding function

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#7625]
**Describe the bug**

In Vim9 script, we cannot evaluate a folding function to make Vim recompute folds.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9script
        sil e /tmp/file
        setl fml=0 fdm=manual fde=getline(v:lnum)=~#'^#'?'>1':'='
        :%d
        sil pu =repeat(['x'], 5)
        :1
        feedkeys('i#', 'nxt')
        setl fdm=expr
        foldlevel(1)
        setl fdm=manual
    EOF
    )

The lines are not folded.

**Expected behavior**

The lines are folded.

**Environment**

 - Vim version: 8.2 Included patches: 1-2305
 - OS: Ubuntu 16.04.7 LTS
 - Terminal: xterm(363)

**Additional context**

If the same code is run in a legacy script, then the lines are correctly folded:
```vim
sil e /tmp/file
setl fml=0 fdm=manual fde=getline(v:lnum)=~#'^#'?'>1':'='
%d
sil pu =repeat(['x'], 5)
1
call feedkeys('i#', 'nxt')
setl fdm=expr
eval foldlevel(1)
setl fdm=manual
```
---

I know about `zx` but it does not preserve manually opened/closed folds.

---

Workaround: evaluate the folding function in the legacy context:
```vim
vim9script
e /tmp/file
setl fml=0 fdm=manual fde=getline(v:lnum)=~#'^#'?'>1':'='
:%d
pu =repeat(['x'], 5)
:1
feedkeys('i#', 'nxt')
setl fdm=expr
fu Func()
    eval foldlevel(1)
endfu
Func()
setl fdm=manual
```

**Comment by lacygoill** [comment]
Ah, it's a regression.  I'll try to find out from where it was introduced.

**Comment by lacygoill** [comment]
> Ah, it's a regression. I'll try to find out from where it was introduced.

It's probably [8.2.1368](https://github.com/vim/vim/releases/tag/v8.2.1368).

---

If `'debug'` is set to throw, then `E1004` is raised:

    E1004: White space required before and after '=~' at "=~#'^#'?'>1':'='"

This is because operators like `=~` must be surrounded by whitespace.

So I fixed the issue like this:

    setl fml=0 fdm=manual fde=getline(v:lnum)\ =~#\ '^#'\ ?\ '>1'\ :\ '='
                                             ^^   ^^    ^^ ^^    ^^ ^^

It's weird that I didn't notice this issue earlier.  Anyway, sorry for the noise.

**Comment by lacygoill** [comment]
I re-open because I think the rule should not be enforced if the option was set in a legacy Vim script.  It should be enforced only if it was set in the Vim9 script context.

For example, here, the lines should be folded, even though whitespace is missing around `=~` and `?:`, because `'foldexpr'` is set in the legacy function `SetOptions()`:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9script
        e /tmp/file
        fu SetOptions()
            setl fml=0 fdm=manual fde=getline(v:lnum)=~#'^#'?'>1':'='
        endfu
        SetOptions()
        :%d
        pu =repeat(['x'], 5)
        :1
        feedkeys('i#', 'nxt')
        setl fdm=expr
        foldlevel(1)
        setl fdm=manual
    EOF
    )

**Comment by lacygoill** [comment]
Also, I don't think that's possible, but could Vim raise an error when we forget the whitespace while setting an `'*expr'` option in a Vim9 script?
```vim
vim9script
setl fde=getline(v:lnum)=~#'^#'?'>1':'='
```
    no error is raised

**Comment by lacygoill** [comment]
> Also, I don't think that's possible, but could Vim raise an error when we forget the whitespace while setting an `'*expr'` option in a Vim9 script?

Ah, I think it's on the todo list:

https://github.com/vim/vim/blob/b23279d7a2d28de5df942924b77cf23672fc684f/runtime/doc/todo.txt#L152-L154

**Comment by lacygoill** [comment]
> It should be enforced only if it was set in the Vim9 script context.

In fact, I wonder whether it should be enforced at all for an `'*expr'` option.  These options typically include operators which are not surrounded by whitespace because of `:h option-backslash`.  Would it make sense to relax the rule for when we set these options?

**Comment by brammool** [comment]
It looks like this happens because the evaluation is done while sourcing a Vim9 script.  That's a bit unexpected.
We should use the information about where the option was set for whether to evaluate it in Vim9 script context.
Compiling it when it's set would catch these errors.  This is not backwards compatible, it may throw an error because something isn't defined yet.  But that is OK for new scripts, and it will catch problems that might otherwise be hard to spot.

**Comment by lacygoill** [comment]
> Compiling it when it's set would catch these errors.

Could we do something similar for custom Ex commands?

For example:
```vim
vim9script
com -range=% Cmd Func(<line1>,<line2>)
def Func(line1: number, line2: number)
    echo line1
    echo line2
enddef
Cmd
```
    E1069: White space required after ','
    E116: Invalid arguments for function Func(1,1)

These errors are confusing because they are raised from the execution of `:Cmd` rather than from its definition.  But the error really comes from the definition; not from the execution:

    com -range=% Cmd Func(<line1>,<line2>)
                                ^^^
                                 ✘

Here, both the execution and the definition are close enough, so it's not too difficult to find the mistake, but it might be much more difficult.  For example, the command could be defined in one script, but call an autoloaded function defined in a completely different script.

Should I open a separate issue, or is it just impossible?

**Comment by lacygoill** [comment]
> For example, the command could be defined in one script, but call an autoloaded function defined in a completely different script.

Sorry, I meant the command could be defined in one script, but be executed in a different script.

**Comment by brammool** [comment]
A user defined command can be anything.  I don't think compiling it is a good idea.
We could make the errors a bit more verbose.

**Comment by lacygoill** [comment]
Fixed by patch [8.2.4173](https://github.com/vim/vim/releases/tag/v8.2.4173), so closing.

Although, it would be nice to give an error earlier, as soon as the option is set, and not wait until it's evaluated to compute the fold level of a line.
